src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -66,9 +66,16 @@ private ITypeSymbol GetSymbolType(ISymbol symbol)
             var matchPriority = preselect ? ComputeSymbolMatchPriority(symbols[0]) : MatchPriority.Default;
             rules = rules.WithMatchPriority(matchPriority);
 
-            if (preselect)
+            if (context.IsSoftSelectionContext)
             {
-                rules = rules.WithSelectionBehavior(PreselectedItemSelectionBehavior);
+                rules = rules.WithSelectionBehavior(CompletionItemSelectionBehavior.SoftSelection);
+            }
+            else
+            {
+                if (preselect)
+                {
+                    rules = rules.WithSelectionBehavior(PreselectedItemSelectionBehavior);
+                }
             }
```

Comment:
somethin feels not great here   can it not be the case that someone can just supply the SelectionBehavior they want?

Reply:
Here are three options of the CompletionItemSelectionBehavior
[https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/Completion/CompletionItemRules.cs#L9](https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/Completion/CompletionItemRules.cs#L9).

Usually, there are preselected items. In the case, they will be hard selected: [https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs#L145](https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs#L145)

We explicitly does not want any item to be hard selected in case of typing a single dot after a numeric. @cyrusnajmabadi, do you see a scenario when we still should allow a hard selection in the case?


---
In reply to: [285306982](https://github.com/dotnet/roslyn/pull/35749#discussion_r285306982) [](ancestors = 285306982)

Reply:
ok... so looking at thsi, i dont' like it, but i think it's a fallout from history (and so i don't thnk it's on you to have to change or make things better).  Namely, we have an impedence mismatch betwene how we used to represent things (i.e. "soft select" or "preselect") and how our current final representation works (i.e. "selection behavior").  This feels like an attempt to munge over these different styles, rather than just updating the providers to use the new approach.  This was likely convenient, but it just means wonkiness like this.

So, i'm fine with this staying as i think "the right thing" here is likely pretty large and complex to do.

Reply:
Thank you for explanations, @cyrusnajmabadi!
I'll take this into account and see how we can improve this when get a better understanding of how it should be arranged. Please do not hesitate to share any advise on this.

---
In reply to: [287161564](https://github.com/dotnet/roslyn/pull/35749#discussion_r287161564) [](ancestors = 287161564)